### DISCUSSION TITLE
Incorrect number of data points in output file when MatrixWriter and CSVWriter both used

### DISCUSSION THREAD

**Discussion by cameycalzolano** [bonsai-rx/bonsai#1241] [Q&A]
Please see the attached workflow.  This is a simplified version of our desired set up where we will be recording from 3 cameras and 3 analog channels from a measurement computing DAQ. I was originally trying to use CombineLatest to align these data streams, but was getting empty files so I've backed up to make sure the input nodes are giving the expected output and to see whether we can just post-hoc align based on timestamps.

We are trying to record analog input from a measurement computing DAQ using the mccdaq AnalogInput node as well as associated timestamps using the TimeStamp node. If we record only the output of AnalogInput with MatrixWriter, we get approximately the expected number of values. If we connect a Timestamp node and record its output with CSVWriter (as shown in the workflow) both outputs are wrong: there's way fewer samples than expected and many timestamps are repeated.

(The other part of the workflow where we are recording from a PS3eye camera  generates timestamps as expected)

Has anyone else ever run into an issue similar to this?

As always, any input is much appreciated. Thank you!
![WorkflowSnapshot](https://user-images.githubusercontent.com/114766484/216472565-9e747aa6-0db7-4a16-bd24-3b54b10d6d4f.jpg)

**Comment by bruno-f-cruz** [comment]
Heya, 

Can you include the workflow? 
thanks

**Reply by cameycalzolano** [reply]
[TimestampTest1.zip](https://github.com/bonsai-rx/bonsai/files/10580884/TimestampTest1.zip)

Thank you for the fast response! Here is the workflow (GitHub wouldn't accept the .bonsai file so I put it in a .zip file). Let me know if I can provide any other information that would be useful, and thank you so much for your help!

**Reply by bruno-f-cruz** [reply]
Repeated timestamps are probably to be expected since you set the buffer size to 1. Try to set this value to 100. Since windows is not an RTOS certain timers are only updated so often. The missing values I am not sure I understand what they are missing in relation to. If you are closing Bonsai without waiting for all the Sinks to close, it could be that MatrixWriter has time to write more samples than the .csv. How are you stopping the workflow?

In general, if you want to align cameras without digital input capability, you should probably just use an LED that blinks in all of them at the same time and record this signal in the NIDAQ too. Relying on system-level timestamping is not that accurate.

**Reply by cameycalzolano** [reply]
Hi Bruno, thank you so much for the reply - here are my responses to your questions:

_Repeated timestamps are probably to be expected since you set the buffer size to 1. Try to set this value to 100._
We tried with a larger buffer as well and had the same problem (repeated timestamps and many fewer timestamps than expected based on the acquisition rate)

_The missing values I am not sure I understand what they are missing in relation to_
We run the workflow with the due time in the Timer node set to 30 seconds. With the sample rate of the AnalogInput node being set to 100 and buffer size 100, we expect 3000 values in the MatrixWriter output and I guess 300 in the CSVWriter output. Even if the timer isn’t updating consistently, shouldn’t the MatrixWriter output the correct amount of data (3000 samples)? We’re getting ~40 values in the .bin file.  
 
_If you are closing Bonsai without waiting for all the Sinks to close, it could be that MatrixWriter has time to write more samples than the .csv. How are you stopping the workflow?_
We are stopping the workflow with a timer (via connecting the Timer & WorkflowOutput nodes at the bottom of the workflow). 

We were originally trying to use CombineLatest to align the camera and mcDAQ input (as suggested in other posts, such as the workflow Goncalo suggests in[ this post ](https://groups.google.com/g/bonsai-users/c/mqfcaZsI6LQ/m/CpkZrolEBQAJ)), but were getting similarly small output files and couldn’t interpret the data in the csvfile (CombineLatest->CSVWriter). If using the output of TimeStamps is inaccurate, is this an alternative approach? I attached a workflow that uses this approach but we’re not getting the expected outputs here either (~40 values instead of 3000 from the MatrixWriter)  

[TimestampTest2.zip](https://github.com/bonsai-rx/bonsai/files/10581831/TimestampTest2.zip)

*Also, note that we are using Bonsai32 as it is the only version of Bonsai compatible with the driver (CL eye multicam driver) that we will be using for our cameras (PS3Eye Camera).

Thanks again for all of your help!
